components/camera/videosource/webcam.go

```diff
@@ -84,137 +63,68 @@ func (c WebcamConfig) Validate(path string) ([]string, []string, error) {
-			} else {
-				constraint.Height = prop.IntRanged{Min: minResolutionDimension, Ideal: 480, Max: 2160}
-			}
-
-			if conf.FrameRate > 0.0 {
-				constraint.FrameRate = prop.FloatExact(conf.FrameRate)
-			} else {
-				constraint.FrameRate = prop.FloatRanged{Min: 0.0, Ideal: 30.0, Max: 140.0}
-			}
-
-			if conf.Format == "" {
-				constraint.FrameFormat = prop.FrameFormatOneOf{
-					frame.FormatI420,
-					frame.FormatI444,
-					frame.FormatYUY2,
-					frame.FormatUYVY,
-					frame.FormatRGBA,
-					frame.FormatMJPEG,
-					frame.FormatNV12,
-					frame.FormatNV21,
-					frame.FormatZ16,
-				}
-			} else {
-				constraint.FrameFormat = prop.FrameFormatExact(conf.Format)
-			}
-
-			logger.Debugf("constraints: %v", constraint)
-		},
-	}
-}
-
-// findReaderAndDriver finds a video device and returns an image reader and the driver instance,
-// as well as the path to the driver.
-func findReaderAndDriver(
-	conf *WebcamConfig,
-	path string,
-	logger logging.Logger,
-) (video.Reader, driverutils.Driver, string, error) {
-	mediadevicescamera.Initialize()
-	constraints := makeConstraints(conf, logger)
-
-	// Handle specific path
-	if path != "" {
-		resolvedPath, err := filepath.EvalSymlinks(path)
-		if err == nil {
-			path = resolvedPath
-		}
-		reader, driver, err := getReaderAndDriver(filepath.Base(path), constraints, logger)
-		if err != nil {
-			return nil, nil, "", err
-		}
-
-		img, release, err := reader.Read()
-		if release != nil {
-			defer release()
-		}
-		if err != nil {
-			return nil, nil, "", err
-		}
-
-		if conf.Width != 0 && conf.Height != 0 {
-			if img.Bounds().Dx() != conf.Width || img.Bounds().Dy() != conf.Height {
-				return nil, nil, "", errors.Errorf("requested width and height (%dx%d) are not available for this webcam"+
-					" (closest driver found supports resolution %dx%d)",
-					conf.Width, conf.Height, img.Bounds().Dx(), img.Bounds().Dy())
-			}
-		}
-		return reader, driver, path, nil
-	}
-
-	// Handle "any" path
-	reader, driver, err := getReaderAndDriver("", constraints, logger)
-	if err != nil {
-		return nil, nil, "", errors.Wrap(err, "found no webcams")
-	}
-	labels := strings.Split(driver.Info().Label, mediadevicescamera.LabelSeparator)
-	if len(labels) == 0 {
-		logger.Error("no labels parsed from driver")
-		return nil, nil, "", nil
-	}
-	path = labels[0] // path is always the first element
-
-	return reader, driver, path, nil
-}
-
-// webcam is a video driver wrapper camera that ensures its underlying driver stays connected.
+// webcam is a video driver wrapper camera that ensures its underlying driver stays connected,
+// handling hot unplugs/replugs, and provides a buffer to read frames from.
 type webcam struct {
 	resource.Named
-	mu                      sync.RWMutex
-	hasLoggedIntrinsicsInfo bool
+	resource.AlwaysRebuild
+	// mu protects all struct fields except workers.
+	//
+	// Lock ordering rules:
+	//   - Never hold mu when calling workers.Stop() or workers.Add()
```

Comment:
Is this deadlock issue something we should look into in other modules using stoppable workers?